arwen/contexts/metering.go

```diff
@@ -78,38 +76,52 @@ func (context *meteringContext) BoundGasLimit(value int64) uint64 {
 	return limit
 }
 
-// DeductAndLockGasIfAsyncStep will deduct the gas for an async step and also lock gas for the callback, if the execution is an asynchronous call
-func (context *meteringContext) DeductAndLockGasIfAsyncStep() error {
-	context.gasLockedForAsyncStep = 0
-
-	input := context.host.Runtime().GetVMInput()
-	if input.CallType != vmcommon.AsynchronousCall {
-		return nil
-	}
-
+// UseGasForAsyncStep consumes the AsyncCallStep gas cost on the currently
+// running Wasmer instance
+// TODO the warmInstance will be affected if this function is misused
+func (context *meteringContext) UseGasForAsyncStep() error {
 	gasSchedule := context.GasSchedule().ElrondAPICost
+	gasToDeduct := gasSchedule.AsyncCallStep
+	return context.UseGasBounded(gasToDeduct)
+}
 
-	gasToLock := gasSchedule.AsyncCallStep + gasSchedule.AsyncCallbackGasLock
-	gasToDeduct := gasSchedule.AsyncCallStep + gasToLock
-	if input.GasProvided <= gasToDeduct {
+func (context *meteringContext) UseGasBounded(gasToUse uint64) error {
+	if context.GasLeft() <= gasToUse {
 		return arwen.ErrNotEnoughGas
 	}
-	input.GasProvided -= gasToDeduct
+	context.UseGas(gasToUse)
+	return nil
+}
 
-	context.gasLockedForAsyncStep = gasToLock
+// ComputeGasToLockForAsync calculates the minimum amount of gas to lock for async callbacks
+func (context *meteringContext) ComputeGasLockedForAsync() uint64 {
+	baseGasSchedule := context.GasSchedule().BaseOperationCost
+	apiGasSchedule := context.GasSchedule().ElrondAPICost
+	codeSize := context.host.Runtime().GetSCCodeSize()
 
-	return nil
+	// Exact amount of gas required to compile this SC again, to execute the
+	// callback
```

Comment:
move this to the end of the previous line

Reply:
Done.